docs/design/extending-api.md

```diff
@@ -0,0 +1,169 @@
+
+
+## Object Model
+
+### The Schema Object
+The `Schema` object describes the object schema(s) of multiple versions of a custom object that the user wants to add
+to the Kubernetes API.
+
+Each Schema object has the standard Kubernetes metadata.
+
+Additionally it has an array of one more `Version` objects.
+
+### The Version Object
+The `Version` object describes a single concrete version of a custom object.  The `Version` object has
+two fields:
+   * `Name` which is the name of the version (e.g. `v1beta3`)
+   * `Spec` which is the specification of the version
+
+### The SchemaSpec Object
+The `SchemaSpec` object describes a concrete schema for an object, it consists of one or more `Field` objects.
+
+
+### The Field Object
+Each `Field` object has two fields:
+   * `Name` which is the name of the field (e.g. `color`)
+   * `Type` which describes the type of the field (e.g. `int`)
+
+### The Type Object
+Each `Type` object has three fields:
+   * `Kind` which specifies the kind of the field.  The possible values correspond to JSON types and are:
+      * `Number`
+      * `String`
+      * `Boolean`
+      * `Array`
+      * `Object`
+   * `Schema` is a required field if the `Kind` is `Object` and may not be specified if `Kind` is not `Object`.
+Schema specifies the sub-schema for an Object
+   * `Type` is a required field if the `Kind` is `Array` and may not be specified if `Kind` is not `Array`.
+
+
+## Usage
+
+When a user creates a new `Schema` object, the Kubernetes API Server reacts by creating a new, namespaced
+RESTful resource path.
+
+For example, if a user creates:
+```yaml
+metadata:
+  name: crontab
+versions:
+  - name: v1
+    fields:
+      - name: cronSpec
+        type: String
+      - name: image
+        type: String
+  - name: v2
+    fields:
+      - name: cronSpec
+        type: String
+      - name: image
+        type: String
+      - name: maxFailures
+        type: Number
+```
+
+Then the API server will program in two new RESTful resource paths:
+   * `/custom/namespaces/<namespace>/crontab/v1/...`
+   * `/custom/namespaces/<namespace>/crontab/v2/...`
+
+
+Now that this schema has been created, a user can `PUT`:
+```json
+{
+   "cronSpec": "* * * * /5",
+   "image":     "my-awesome-chron-image"
+}
+```
+
+to: `/custom/namespaces/default/crontab/v1/my-crontab`
+
+and the corresponding data will be stored into etcd by the APIServer, so that when the user issues:
+
+```
+GET /custom/namespaces/default/crontab/v1/my-contrab
+```
+
+And when they do that, they will get back the same data.
+
+Likewise, to list all resources, a user can issue:
+```
+GET /custom/namespaces/default/crontab/v1/my-contrab
+```
+
+and get back:
+```json
+{
+   "items": [
+     {
+       "cronSpec": "* * * * /5",
+       "image":    "my-awesome-chron-image"
+     }
+   ]
+}
+```
+
+Note that for now `POST` is not supported as a RESTful verb, since it requires a naming
```

Comment:
I agree. We need POST.

We should also support PUT-to-create, but plugins should be able to utilize the same API conventions as built-in API objects.